### PR TITLE
Support OIDC Authentication

### PR DESCRIPTION
**Pull request by manicminer**
Closes: #91

### CODE REVIEW

**Comment by manicminer** [comment]
Test results

Screenshot 2022-04-25 at 21 35 04

File: authentication/auth_method_oidc.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+		ClientID:            a.clientId,
+		IDTokenRequestURL:   a.idTokenRequestUrl,
+		IDTokenRequestToken: a.idTokenRequestToken,
+		Scopes:              []string{api.DefaultScope()},
+	}
+
+	return &authWrapper.Authorizer{Authorizer: conf.TokenSource(ctx)}, nil
+}
+
+func (a oidcAuth) populateConfig(c *Config) error {
```
Comment (tombuildsstuff) [inline authentication/auth_method_oidc.go:65]: this'll need to be a pointer:

```suggestion
func (a *oidcAuth) populateConfig(c *Config) error {
```

File: authentication/auth_method_oidc_test.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+func TestOIDC_populateConfig(t *testing.T) {
+	config := &Config{}
+	err := oidcAuth{}.populateConfig(config)
+	if err != nil {
+		t.Fatalf("Error populating config: %s", err)
+	}
+	if !config.AuthenticatedAsAServicePrincipal {
+		t.Fatalf("Expected `AuthenticatedAsAServicePrincipal` to be true but it wasn't")
+	}
```
Comment (tombuildsstuff) [inline authentication/auth_method_oidc_test.go:89]: given we aren't guaranteed a refresh token when using OIDC, I think we could do with a flag to know this is OIDC to handle that under certain conditions?

**Review by tombuildsstuff — APPROVED** [review]
Left a few comments inline but otherwise 👍 nice one

File: authentication/auth_method_oidc_test.go (hunk 2 above)
Comment (manicminer) [reply authentication/auth_method_oidc_test.go:89]: We automatically get a new ID token from GitHub and use that to get a new access token to essentially fake a refresh - if that fails then there are essentially no recovery options to use anyway.

**Pull request merged**